### PR TITLE
Remove contention on worker.currentMu by maintaining a snapshot copy of pending Block and StateDB.

### PR DESCRIPTION
**Pull request by ryanschneider**
Related to #15859, this change removes contention on the `worker.currentMu` mutex for nodes serving a lot of RPC traffic.

In my testing, vanilla 1.8.3 can easily take 2-3 seconds to respond to `eth_estimateGas` when under load, while this version routinely takes ~50ms.

The contention is removed by updating "snapshot" Block and StateDB entries, and returning those in `pending()` and `pendingBlock()`, except when mining, since that seems to want to return the actual block rather than a copy of it.

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 6a687b90e01a67df572c0e6d72fa2406d55c9bd6
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: miner/worker.go [hunk 1]
```diff
@@ -171,32 +175,28 @@ func (self *worker) setExtra(extra []byte) {
...
-		return types.NewBlock(
-			self.current.header,
-			self.current.txs,
-			nil,
-			self.current.receipts,
-		), self.current.state.Copy()
+		// return a snapshot to avoid contention on currentMu mutex
+		self.snapshotMu.RLock()
+		defer self.snapshotMu.RUnlock()
+		return self.snapshotBlock, self.snapshotState.Copy()
```
Comment (ryanschneider) [inline miner/worker.go]: NOTE: Here we are returning the *same* snapshotBlock to all consumers, rather than a copy each time.  That seemed ok to me in my testing, but we could be more conservative and use `types.NewBlock(...)` here to make another copy of the `snapshotBlock`.  Same applies for `pendingBlock()` below.

**Comment by ryanschneider** [comment]
I'll update the commit message with the package name, and go ahead and use `NewBlock` to be completely safe.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 54a35832967b2021a16cec7570cb15e5973e28b5
  * Commit subjects should be kept under 100 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by ryanschneider** [comment]
Ok, commit message is in approved format now.  I'm going to hold off on the NewBlock change though until I get some feedback on if that's necessary, as what I really need is a `CopyBlock(b *Block) *Block` function but I don't see something like that at the moment, and only want to add it if necessary.

File: miner/worker.go (hunk 1 above)
Comment (karalabe) [reply miner/worker.go]: A block is immutable AFAIK, so it should be safe to return the same thing for everyone.

**Comment by karalabe** [comment]
Out of curiosity, if the contention is around the mutex, wouldn't it make more sense to replace it with a RWMutex and only RLock it for the pending state retrievals? That might be a simpler solution if indeed it's a lock contention.

**Comment by ryanschneider** [comment]
The problem is that the current `currentMu.Lock()` calls in `worker.commitNewWork()` and `worker.update()` (on receiving a TxPreEvent) are both very coarse, so the lock is held for a long time.  I'm not familiar enough with the code to try to chop up those locks, especially in `commitNewWork`, which aliases `work := self.current` and then passes both `work` and `self.current` to other systems.

Digging deeper, I feel like the `work` variable introduced in `commitNewWork` should actually be  a deep copy, since it's passed into the  channel(s) returned by`Agent.SetReturnCh()` for all the agents registered via `Miner.Register(a *Agent)`.  If this was the case, then it *might* be possible to replace `currentMu` with a RWMutex inside `Work.commitTransactions`, but I think a bit more shuffling would still be needed.

In short, I think making the locks in `worker.go` more fine-grained is pretty hairy and outside of my familiarity with the codebase, hence the snapshot work-around.

**Comment by ryanschneider** [comment]
I forgot to say that I'm all for making this the best solution, so if I get time to dig into `worker.go` more I will, and if you can provide any guidance I'm happy to help chop up the `currentMu` more.

I'll try to get some pprof dumps of vanilla 1.8.3 responding slowly to `eth_estimateGas` so we have more data to work with.

**Comment by ryanschneider** [comment]
Here's a 30 second mutex profile from the code in question in vanilla geth 1.8.3:

```
File: geth-linux-amd64
Type: delay
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 14517.07ms, 70.54% of 20580.50ms total
Dropped 2 nodes (cum <= 102.90ms)
      flat  flat%   sum%        cum   cum%
 8801.64ms 42.77% 42.77%  8801.64ms 42.77%  github.com/ethereum/go-ethereum/miner.(*worker).pending /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
 3217.63ms 15.63% 58.40%  5343.29ms 25.96%  github.com/ethereum/go-ethereum/miner.(*worker).update /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
 2053.62ms  9.98% 68.38%  2087.68ms 10.14%  github.com/ethereum/go-ethereum/miner.(*worker).commitNewWork /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
  444.17ms  2.16% 70.54%   444.17ms  2.16%  github.com/ethereum/go-ethereum/miner.(*worker).pendingBlock /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
         0     0% 70.54%  8801.64ms 42.77%  github.com/ethereum/go-ethereum/miner.(*Miner).Pending /ext-go/1/src/github.com/ethereum/go-ethereum/miner/miner.go
         0     0% 70.54%   444.17ms  2.16%  github.com/ethereum/go-ethereum/miner.(*Miner).PendingBlock /ext-go/1/src/github.com/ethereum/go-ethereum/miner/miner.go
```

And a 30 sample from the code in this PR:

```
File: geth-linux-amd64
Type: delay
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 119.45ms, 0.98% of 12176.66ms total
      flat  flat%   sum%        cum   cum%
  112.85ms  0.93%  0.93%   112.85ms  0.93%  github.com/ethereum/go-ethereum/miner.(*Work).commitTransaction /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
    6.60ms 0.054%  0.98%    68.25ms  0.56%  github.com/ethereum/go-ethereum/miner.(*worker).commitNewWork /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
         0     0%  0.98%   112.76ms  0.93%  github.com/ethereum/go-ethereum/miner.(*Work).commitTransactions /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
         0     0%  0.98%   112.73ms  0.93%  github.com/ethereum/go-ethereum/miner.(*worker).update /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
```

**Comment by ryanschneider** [comment]
^ Sorry I hit "Close and comment" instead of "Comment" :) 

My interpretation of the above mutex profile is that in vanilla 1.8.3, ~40% of the time `worker.pending()` is blocked on the `currentMu` mutex, while with my patch it's basically never blocked.

I've also included block profile output below, but honestly find it much harder to infer anything from.

1.8.3:

```
Showing nodes accounting for 1.90mins, 3.25% of 58.48mins total
Dropped 3 nodes (cum <= 0.29mins)
      flat  flat%   sum%        cum   cum%
  1.52mins  2.59%  2.59%   1.56mins  2.67%  github.com/ethereum/go-ethereum/miner.(*worker).update /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
  0.24mins  0.42%  3.01%   0.24mins  0.42%  github.com/ethereum/go-ethereum/miner.(*Miner).update /ext-go/1/src/github.com/ethereum/go-ethereum/miner/miner.go
  0.09mins  0.16%  3.17%   0.09mins  0.16%  github.com/ethereum/go-ethereum/miner.(*worker).pending /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
  0.03mins 0.046%  3.22%   0.03mins 0.048%  github.com/ethereum/go-ethereum/miner.(*worker).commitNewWork /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
  0.02mins 0.033%  3.25%   0.02mins 0.033%  github.com/ethereum/go-ethereum/miner.(*Work).commitTransaction /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
         0     0%  3.25%   0.09mins  0.16%  github.com/ethereum/go-ethereum/miner.(*Miner).Pending /ext-go/1/src/github.com/ethereum/go-ethereum/miner/miner.go
         0     0%  3.25%   0.02mins 0.033%  github.com/ethereum/go-ethereum/miner.(*Work).commitTransactions /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
```

Patched:

```
(pprof) top
Showing nodes accounting for 2.69mins, 2.56% of 105.20mins total
Dropped 3 nodes (cum <= 0.53mins)
      flat  flat%   sum%        cum   cum%
  1.44mins  1.37%  1.37%   1.48mins  1.40%  github.com/ethereum/go-ethereum/miner.(*worker).update /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
  1.19mins  1.13%  2.50%   1.19mins  1.13%  github.com/ethereum/go-ethereum/miner.(*Miner).update /ext-go/1/src/github.com/ethereum/go-ethereum/miner/miner.go
  0.02mins 0.023%  2.52%   0.02mins 0.023%  github.com/ethereum/go-ethereum/miner.(*worker).pending /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
  0.02mins  0.02%  2.54%   0.02mins  0.02%  github.com/ethereum/go-ethereum/miner.(*worker).commitNewWork /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
  0.02mins 0.015%  2.56%   0.02mins 0.015%  github.com/ethereum/go-ethereum/miner.(*Work).commitTransaction /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
         0     0%  2.56%   0.02mins 0.023%  github.com/ethereum/go-ethereum/miner.(*Miner).Pending /ext-go/1/src/github.com/ethereum/go-ethereum/miner/miner.go
         0     0%  2.56%   0.02mins 0.015%  github.com/ethereum/go-ethereum/miner.(*Work).commitTransactions /ext-go/1/src/github.com/ethereum/go-ethereum/miner/worker.go
```

**Comment by karalabe** [comment]
There seems to be a small data race in your code. Maybe I'm wrong, feel free to correct me.

The snapshots are only initialized when work is first submitted to the miner. But if I quickly create the miner and retrieve the pending block/state, I'll either return `nil`, or crash on the `Copy`. Might be worthwhile to have an extra check that if the snapshots are `nil`, then the actual real pending stuff is copied.

**Comment by karalabe** [comment]
Truth be told, if there's a race in your code, there is probably one in the original code too. As such, lets not worry about it now.
